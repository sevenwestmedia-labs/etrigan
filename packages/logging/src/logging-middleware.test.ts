import pino from 'pino'

import { expressRequestLoggingMiddleware } from './'
it('logs requests', async () => {
    const logs: string[] = []
    const events: { [event: string]: () => any } = {}
    const log = pino(
        {
            level: 'debug',
            timestamp: false,
            base: {},
        },
        {
            write(msg) {
                logs.push(msg)
            },
        },
    )
    const middleware = expressRequestLoggingMiddleware(log, {
        createRequestId() {
            return '<requestid>'
        },
        now() {
            return 12345
        },
    })
    const res = {
        on(event: string, cb: () => any) {
            events[event] = cb.bind(res)
        },
        removeListener(event: string) {
            delete events[event]
        },
    }
    middleware(
        {
            header() {
                return undefined
            },
            query: {},
        } as any,
        res as any,
        () => {},
    )
    events.finish()

    expect(logs).toMatchSnapshot()
    expect(Object.keys(events).length).toBe(0)
})

it('logs API caller headers', async () => {
    const logs: string[] = []
    const events: { [event: string]: () => any } = {}
    const log = pino(
        {
            level: 'debug',
            timestamp: false,
            base: {},
        },
        {
            write(msg) {
                logs.push(msg)
            },
        },
    )
    const middleware = expressRequestLoggingMiddleware(log, {
        createRequestId() {
            return '<requestid>'
        },
        now() {
            return 12345
        },
    })
    const res = {
        on(event: string, cb: () => any) {
            events[event] = cb.bind(res)
        },
        removeListener(event: string) {
            delete events[event]
        },
    }
    middleware(
        {
            header(name: string) {
                if (name === 'x-request-id') {
                    return 'testRequest1'
                }
                if (name === 'Caller') {
                    return 'custom API caller'
                }
                return undefined
            },
            query: {},
        } as any,
        res as any,
        () => {},
    )
    events.finish()

    expect(logs).toMatchSnapshot()
    expect(Object.keys(events).length).toBe(0)
})

it('logs failed', async () => {
    const logs: string[] = []
    const events: { [event: string]: () => any } = {}
    const log = pino(
        {
            level: 'debug',
            timestamp: false,
            base: {},
        },
        {
            write(msg) {
                logs.push(msg)
            },
        },
    )
    const middleware = expressRequestLoggingMiddleware(log, {
        createRequestId() {
            return '<requestid>'
        },
        now() {
            return 12345
        },
    })
    const res = {
        on(event: string, cb: () => any) {
            events[event] = cb.bind(res)
        },
        removeListener(event: string) {
            delete events[event]
        },
    }
    middleware(
        {
            header() {
                return undefined
            },
            query: {},
        } as any,
        res as any,
        () => {},
    )
    events.error()

    expect(logs).toMatchSnapshot()
    expect(Object.keys(events).length).toBe(0)
})

it('logs close', async () => {
    const logs: string[] = []
    const events: { [event: string]: () => any } = {}
    const log = pino(
        {
            level: 'debug',
            timestamp: false,
            base: {},
        },
        {
            write(msg) {
                logs.push(msg)
            },
        },
    )
    const middleware = expressRequestLoggingMiddleware(log, {
        createRequestId() {
            return '<requestid>'
        },
        now() {
            return 12345
        },
    })
    const res = {
        on(event: string, cb: () => any) {
            events[event] = cb.bind(res)
        },
        removeListener(event: string) {
            delete events[event]
        },
    }
    middleware(
        {
            header() {
                return undefined
            },
            query: {},
        } as any,
        res as any,
        () => {},
    )
    events.close()

    expect(logs).toMatchSnapshot()
    expect(Object.keys(events).length).toBe(0)
})
