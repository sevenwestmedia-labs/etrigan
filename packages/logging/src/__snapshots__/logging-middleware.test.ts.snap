// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logs API caller headers 1`] = `
Array [
  "{\\"level\\":20,\\"requestId\\":\\"testRequest1\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"testRequest1\\",\\"caller\\":\\"custom API caller\\",\\"log\\":{},\\"startTime\\":12345},\\"msg\\":\\"begin request\\"}
",
  "{\\"level\\":30,\\"requestId\\":\\"testRequest1\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"testRequest1\\",\\"caller\\":\\"custom API caller\\",\\"log\\":{},\\"startTime\\":12345},\\"res\\":{\\"req\\":{\\"query\\":{},\\"requestId\\":\\"testRequest1\\",\\"caller\\":\\"custom API caller\\",\\"log\\":{},\\"startTime\\":12345},\\"endTime\\":12345},\\"responseTime\\":0,\\"msg\\":\\"end request\\"}
",
]
`;

exports[`logs close 1`] = `
Array [
  "{\\"level\\":20,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"msg\\":\\"begin request\\"}
",
  "{\\"level\\":30,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"res\\":{\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"endTime\\":12345},\\"responseTime\\":0,\\"msg\\":\\"end request - connection closed\\"}
",
]
`;

exports[`logs failed 1`] = `
Array [
  "{\\"level\\":20,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"msg\\":\\"begin request\\"}
",
  "{\\"level\\":30,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"res\\":{\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"endTime\\":12345},\\"responseTime\\":0,\\"msg\\":\\"end request\\"}
",
]
`;

exports[`logs requests 1`] = `
Array [
  "{\\"level\\":20,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"msg\\":\\"begin request\\"}
",
  "{\\"level\\":30,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"res\\":{\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"endTime\\":12345},\\"responseTime\\":0,\\"msg\\":\\"end request\\"}
",
]
`;
